scene/gui/control.cpp

```diff
@@ -2978,35 +2977,28 @@ Control *Control::make_custom_tooltip(const String &p_text) const {
 // Base object overrides.
 
 void Control::add_child_notify(Node *p_child) {
-	Control *child_c = Object::cast_to<Control>(p_child);
-
-	if (child_c && child_c->data.theme.is_null() && (data.theme_owner || data.theme_owner_window)) {
-		_propagate_theme_changed(child_c, data.theme_owner, data.theme_owner_window); //need to propagate here, since many controls may require setting up stuff
-	}
-
-	Window *child_w = Object::cast_to<Window>(p_child);
-
-	if (child_w && child_w->theme.is_null() && (data.theme_owner || data.theme_owner_window)) {
-		_propagate_theme_changed(child_w, data.theme_owner, data.theme_owner_window); //need to propagate here, since many controls may require setting up stuff
+	// We propagate when this node uses a custom theme, so it can pass it on to its children.
+	if (data.theme_owner || data.theme_owner_window) {
+		// `p_notify` is false here as `NOTIFICATION_THEME_CHANGED` will be handled by `NOTIFICATION_ENTER_TREE`.
+		_propagate_theme_changed(p_child, data.theme_owner, data.theme_owner_window, false, true);
 	}
 }
 
 void Control::remove_child_notify(Node *p_child) {
-	Control *child_c = Object::cast_to<Control>(p_child);
-
-	if (child_c && (child_c->data.theme_owner || child_c->data.theme_owner_window) && child_c->data.theme.is_null()) {
-		_propagate_theme_changed(child_c, nullptr, nullptr);
-	}
-
-	Window *child_w = Object::cast_to<Window>(p_child);
-
-	if (child_w && (child_w->theme_owner || child_w->theme_owner_window) && child_w->theme.is_null()) {
-		_propagate_theme_changed(child_w, nullptr, nullptr);
+	// If the removed child isn't inheriting any theme items through this node, then there's no need to propagate.
+	if (data.theme_owner || data.theme_owner_window) {
+		_propagate_theme_changed(p_child, nullptr, nullptr, false, true);
 	}
 }
 
 void Control::_notification(int p_notification) {
 	switch (p_notification) {
+		case NOTIFICATION_ENTER_TREE: {
+			// Need to defer here, because theme owner information might be set in
+			// add_child_notify, which doesn't get called until right after this.
+			call_deferred("notification", NOTIFICATION_THEME_CHANGED);
```

Comment:
```suggestion
			call_deferred(SNAME("notification"), NOTIFICATION_THEME_CHANGED);
```
This is going to be called a lot and the function takes a StringName.
Same below.

Reply:
Great call.